enhancements/baremetal/continuous-hw-introspection.md

```diff
@@ -0,0 +1,147 @@
+
+This proposal talks about how performing continuous hardware introspection on all master
+and worker nodes in a Baremetal IPI deployment would get us important data about baremetal
+hosts which can be used to monitor changes in hardware capabilities of the server.
+
+This enhancement will not result in a new CRD or API to be written but will detail
+data that would become availabe in the Status field of the BareMetalHost CRD.
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [X] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+Continuous hardware inspection on all the master and worker nodes that are a part
+of a Baremetal IPI deployment, allow the user to detect changes in server hardware
+and become aware of any failures seen on the servers. This information can be used
+to report current status of hardware to the user and/or glean metrics about the
+health of the deployment.
+
+At the end of 4.4, a Baremetal IPI deployment does not have continuous hardware introspection
+running on its master or worker nodes. The underlying provisioning service already has the
+capability to perform hardware inspection on each of the servers periodically. The
+baremetal-operator (BMO) then reads this information and updates the "status" field of the
+BareMetalHost CRD. Please refer to [1] for the latest BaremetalHost CRD.
+
+## Motivation
+
+This enhancement will allow the user to get continuous updates on all the baremetal
+servers in their Metal3 deployment providing useful information about the hardware
+capabilities of the servers.
+
+### Goals
+
+This enhancement will allow the baremetal host CR associated with each baremetal
+server in a metal3 deployment to contain up-to-date informtaion about its hardware and
+operational status of a server.
+
+### Non-Goals
+
+This proposal does not include any changes to the BareMetalHost CRD.
+
+## Proposal
+
+This enhancement proposes to add a new DaemonSet started by the machine-api-operator
+only for the BareMetal platform. This DaemonSet would run a new container image
+(ironic-hardware-inventory-recorder) on all master and worker nodes. With the new
+addition, it would become possible to have detailed information regarding the
+hardware connected to both master and worker nodes in a metal3 deployment and not
+just when the worker nodes are provisioned. This also allows the user to keep
+track of any changes in hardware capabilities of servers in a metal3 deployment.
+
+Additional details regarding enhancements made to the underlying service required
+for this feature can be found in [3].
+
+### User Stories [optional]
+
+#### Story 1
+
+As a user, I would like to know the current hardware and its capabilities for each
```

Comment:
How does this relate to the node-feature-discovery efforts? Can you hook into those?
- https://github.com/kubernetes-sigs/node-feature-discovery#local-user-specific-features 
- https://github.com/openshift/cluster-nfd-operator

Reply:
A major difference between this and other kube-centric functionality is that we actually run these tools first before RHCOS is even deployed so we can gather hardware details about all the hosts in our inventory, even before they have been provisioned to be part of a cluster.  This enhancement is how to continue running those tools later to ensure the data we have stays up to date.

It's really a minor update to something already being used, and switching to alternative tooling would be a much broader effort, including figuring out how to run these tools to collect data in a ramdisk, without the host yet being a kubernetes Node.

Some consolidation around hardware inspection tooling sounds like a good long term target, but not so pragmatic given current architecture and needs.

Reply:
It seems to me the purpose of this proposal is to discover host capabilities so we can expose that information to workloads.  We should try to conform with the upstream way of doing this.  A combination of NFD and device-plugins seems to be the right course.  If we install CoreOS on actual bare metal (eg, not using metal3), then that is what I think we would require.  We should try to be as consistent as possible across our deployment types.

Reply:
This feature allows us to report on parameters like model number, vendor firmware version etc of the hardware plugged into the server. (see https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_08_baremetalhost.crd.yaml#L242). 
So, this is different from the data reported by the NFD. But, having said that we would need to add support for the NFD also for the Baremetal IPI platform.

Reply:
The service that performs continuous hardware introspection collects information that is described here (https://github.com/openshift/machine-api-operator/blob/master/install/0000_30_machine-api-operator_08_baremetalhost.crd.yaml#L242). Our concern regarding augmenting NFD to report these values is that some of this information might not be available for other platform types. 
In any case, we understand the need to collaborate with the NFD team. We are working with our PM to prioritize this effort based on customer feedback from this feature. For those reasons, we would like to proceed with the current approach detailed in this proposal.